### ISSUE TITLE
Server to test libp2p `unstable` branch

### ISSUE THREAD

**Issue by dryajov** [status-im/infra-nimbus#43]
libp2p needs to move forward with new features and potentially larger and riskier changes/refactors. In order to prevent disrupting production code that relies on a stable `master` branch, all major changes are now being worked in the `unstable` branch. For further background, take a look at https://github.com/status-im/nim-libp2p/pull/542

We want to replicate the same flow of libp2p `master` for testing with libp2p `unstable`. Essentially, we want a new box where we can deploy the `unstable` branch in the same way we do with `master`, which (AFAIK) is currently being deployed to `unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont` with the `nim-libp2p-auto-bump` branch. Just like with `master`, there will be an auto-bump (`nim-libp2p-auto-bump-unstable`) branch in nimbus-eth for the nim-libp2p `unstable` branch that will be updated every time a new commit lands there. This will need to eventually be migrated to the prater network (https://github.com/status-im/infra-nimbus/issues/41).

To summarize:

- We need an additional box exactly the same as  `unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont`
- A new autobump job will be generated to create a branch to deploy to this new machine 
  - Please let me know if this needs to happen first

**Comment by jakubgs** [comment]
Soooooo, should we rename the one that currently exists to be:
```
stable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
To fit the branch being "stable"?
The new one should be of the same AWS EC2 instance type?

**Comment by dryajov** [comment]
Hrm... we've a conflict of what `stable`/`unstable` is in this context. The stable/unstable in the current names only reflect the nimbus side, but with the introduction of `unstable` in libp2p, this add another level to it. We might name it something like `unstable-main-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont` and `unstable-unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont`, suggestions welcome :)

**Comment by jakubgs** [comment]
```
unstable-unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
Wat?

**Comment by dryajov** [comment]
> ```
> unstable-unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
> ```
> 
> Wat?

LOL, naming is the hardest part of our jobs.

On a serious note, what we need to reflect is that we're testing libp2p `master` or `unstable` against nimbus `unstable`, hence `unstable-unstable` or `master-unstable`, but I know it's a horrible name, so again ideas are welcome.

**Comment by jakubgs** [comment]
Right. So the situation right now is:
```
nim-libp2p-auto-bump branch   -> unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
What you suggested is:
```
nim-libp2p-auto-bump          -> stable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
nim-libp2p-auto-bump-unstable -> untable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
But what would really make sense would be:
```
nim-libp2p-auto-bump-stable   -> stable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
nim-libp2p-auto-bump-unstable -> unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
Also, can you explain to me what the `nim-libp2p-auto-bump` branch even is? And what does `auto-bump` refer to?

**Comment by dryajov** [comment]
> Right. So the situation right now is:
>
> nim-libp2p-auto-bump branch   -> unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont

Yes.

> What you suggested is:
>
> nim-libp2p-auto-bump          -> stable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
> nim-libp2p-auto-bump-unstable -> untable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont

Very close, but more like

```
<nimbus-branch>-<libp/product name (eg. libp2p)>-<lib-branch>-<box-size>-<region>-1a.nimbus.<network>
```

So it would end up looking like this

```
nim-libp2p-auto-bump-stable  -> unstable-libp2p-stable-small-01.aws-eu-central-1a.nimbus.pyrmont
nim-libp2p-auto-bump-unstable -> unstable-libp2p-unstable-small-01.aws-eu-central-1a.nimbus.pyrmont
```

NOTE: The current name of the autobump branch needs to change from `nim-libp2p-auto-bump` to `nim-libp2p-auto-bump-stabe` and we need to rename the `master` branch to `stable` in libp2p

**Comment by dryajov** [comment]
cc @sinkingsugar

**Comment by jakubgs** [comment]
I don't get what kind of information is the first `unstable` communicating in the hostname.

**Comment by dryajov** [comment]
> I don't get what kind of information is the first `unstable` communicating in the hostname.

```
nim-libp2p-auto-bump-stable  -> unstable-libp2p-stable-small-01.aws-eu-central-1a.nimbus.pyrmont
```

Is testing nimbus `unstable` against libp2p `stable`.

```
nim-libp2p-auto-bump-unstable -> unstable-libp2p-unstable-small-01.aws-eu-central-1a.nimbus.pyrmont
```

Is testing nimbus `unstable` against libp2p `unstable`.

In the future we might add more boxes to test nimbus `stable` against the two libp2p branches and it could look like 

```
nim-libp2p-auto-bump-stable    -> stable-libp2p-stable-small-01.aws-eu-central-1a.nimbus.pyrmont
nim-libp2p-auto-bump-unstable  -> stable-libp2p-unstable-small-01.aws-eu-central-1a.nimbus.pyrmont
```

**Comment by sinkingsugar** [comment]
> Also, can you explain to me what the `nim-libp2p-auto-bump` branch even is? And what does `auto-bump` refer to?

Those branches are managed by a bot, that is why "auto".
* Every time a CI build succeeds in the libp2p repo `master` it will open a PR in nimbus `auto-bump` with `unstable` as base
* Every time a CI build succeeds in the libp2p repo `unstable` it will open a PR in nimbus `auto-bump-unstable` with `unstable` as base

We never use the bot to target `stable` nimbus.

**Comment by jakubgs** [comment]
>In the future we might add more boxes to test nimbus stable against the two libp2p branches and it could look like

I see, that makes sense then. Thanks for explaining.

>We never use the bot to target stable nimbus.

Right, but you might in the future.

**Comment by jakubgs** [comment]
Okay, then I think this naming is fine, even though it's quite the mouthful...
```
unstable-libp2p-stable-small-01.aws-eu-central-1a.nimbus.pyrmont
unstable-libp2p-unstable-small-01.aws-eu-central-1a.nimbus.pyrmont
```
But I see the branches have not been renamed yet. I will deploy the host using the current naming and we can adjust later.

**Comment by jakubgs** [comment]
I renamed the old host from `unstable-libp2p-small-01` to `unstable-libp2p-unstable-small-01` in https://github.com/status-im/infra-nimbus/commit/c10f9f57.

**Comment by jakubgs** [comment]
Oh fuck:
```
name cannot be longer than 64 characters on systemd servers, try a shorter name
```
Looks like this won't work...

**Comment by jakubgs** [comment]
I don't think we can make this work. We need some kind of shortcuts or just drop a token.

**Comment by jakubgs** [comment]
The simplest option is to drop the `unstable-` prefix, which will make the hostname under 64 characters.

Another option is some letter coding, like:

* `s` - stable
* `u` - unstable
* `t` - testing

And then use something like:
```
u-libp2p-s-small-01.aws-eu-central-1a.nimbus.pyrmont
u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont
```
or
```
libp2p-u2s-small-01.aws-eu-central-1a.nimbus.pyrmont
libp2p-u2u-small-01.aws-eu-central-1a.nimbus.pyrmont
```
Thoughts?

**Comment by dryajov** [comment]
Lets use this:

```
u-libp2p-s-small-01.aws-eu-central-1a.nimbus.pyrmont
u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont
```

**Comment by jakubgs** [comment]
Wait, so wouldn't the current host we have:
```
unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
```
Would have to be renamed to:
```
s-libp2p-s-01.aws-eu-central-1a.nimbus.pyrmont
```
Since it's `nim-libp2p-auto-bump`(which should be `nim-libp2p-auto-bump-stable`) which is the `master`(which should be `stable`) of the `nim-libp2p` repo merged into `stable` of `nimbus-eth2`.

Am I getting that right? This is confusing.

**Comment by dryajov** [comment]
Yep, let's just setup the boxes to point to all those new jobs/branches
while we change those names on our side in parallel.

**Comment by dryajov** [comment]
@sinkingsugar Can you please do the proper changes to the CI jobs? We can probably leave renaming master for later unless it is uncomplicated.

**Comment by arthurk** [comment]
To clarify is this table correct showing the code branch, release branch and the host it is being deployed to?

| Code Branch | Release Branch | Host
| --- | --- | --- |
| master | nim-libp2p-auto-bump | u-libp2p-s-small-01.aws-eu-central-1a.nimbus.pyrmont
| unstable | nim-libp2p-auto-bump-unstable | u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont

**Comment by dryajov** [comment]
Yes, this is correct.

**Comment by jakubgs** [comment]
Okay, so the steps necessary are:

1. Rename `unstable-libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont` to `u-libp2p-s-small-01.aws-eu-central-1a.nimbus.pyrmont`
2. Deploy a new `u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont` host
3. Setup new builds for the `nim-libp2p-auto-bump-unstable` branch
4. Deploy the `nim-libp2p-auto-bump-unstable` image to `u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont`

@arthurk can you do this since you are already doing Nimbus stuff?

**Comment by jakubgs** [comment]
@arthurk one pro-tip for renaming the host:

When renaming an AWS host Terraform will want to remove the Security Group and create a new one. But a host can't have no security group. The solution to that is to:

1. Use `terraform state rm` to make terraform forget about the current SG
2. Apply changes which will add a new SG to the instance
3. Go into Nimbus AWS console and clean up manually the old SG

Annoying, but it is what it is.

**Comment by arthurk** [comment]
I'm working on it. This should be done for Pyrmont and Prater or is Prater enough?

**Comment by dryajov** [comment]
> I'm working on it. This should be done for Pyrmont and Prater or is Prater enough?

Prater should be enough.

**Comment by jakubgs** [comment]
@arthurk can you update this issue with your progress?

**Comment by arthurk** [comment]
Yes I'll post updates here. I started work on this yesterday and there wasn't much to report.

**Comment by arthurk** [comment]
The stable server is renamed but the unstable server can't run because the image fails to build:

```
Building: Nim compiler
Building: install/usr/lib/libbacktrace.a
Building: install/usr/lib/libbacktracenim.a

Tip of the day: this will probably build faster if you use "make -j2 ...".

Building: build/nimbus_beacon_node
/data/beacon-node-builds/unstable-libp2p-unstable-large/repo/beacon_chain/networking/eth2_network.nim(1541, 44) Error: type mismatch: got <type TcpTransport, set[ServerFlags]>
but expected one of: 
func init(T: type TcpTransport; flags: set[ServerFlags] = {}; upgrade: Upgrade): T:type
  first type mismatch at position: 3
  missing parameter: upgrade
129 other mismatching symbols have been suppressed; compile with --showAllMismatches:on to see them

expression: init(TcpTransport, {ReuseAddr})
make: *** [Makefile:296: nimbus_beacon_node] Error 1
```

**Comment by jakubgs** [comment]
Cool, I see the hosts for Prater are there:
```
 > ansible localhost -m debug -a var='groups["nimbus.prater"]' | grep libp2p
        "u-libp2p-s-large-01.aws-eu-central-1a.nimbus.prater",
        "u-libp2p-u-large-01.aws-eu-central-1a.nimbus.prater",
```
@dryajov can you take a look at the build error?

**Comment by dryajov** [comment]
> The stable server is renamed but the unstable server can't run because the image fails to build:
> 
> ```
> Building: Nim compiler
> Building: install/usr/lib/libbacktrace.a
> Building: install/usr/lib/libbacktracenim.a
> 
> Tip of the day: this will probably build faster if you use "make -j2 ...".
> 
> Building: build/nimbus_beacon_node
> /data/beacon-node-builds/unstable-libp2p-unstable-large/repo/beacon_chain/networking/eth2_network.nim(1541, 44) Error: type mismatch: got <type TcpTransport, set[ServerFlags]>
> but expected one of: 
> func init(T: type TcpTransport; flags: set[ServerFlags] = {}; upgrade: Upgrade): T:type
>   first type mismatch at position: 3
>   missing parameter: upgrade
> 129 other mismatching symbols have been suppressed; compile with --showAllMismatches:on to see them
> 
> expression: init(TcpTransport, {ReuseAddr})
> make: *** [Makefile:296: nimbus_beacon_node] Error 1
> ```

Is this a blocker? It's expected because of changes in `unstable` not being compatible with nimbus `unstable`. This should fix it tho - https://github.com/status-im/nimbus-eth2/pull/2474.

**Comment by jakubgs** [comment]
So this is the current layout:

| nimbus-eth2 | nim-libp2p | Release Branch | Host
| --- | --- | --- | --- |
| unstable | master | nim-libp2p-auto-bump | u-libp2p-s-small-01.aws-eu-central-1a.nimbus.pyrmont
| unstable | unstable | nim-libp2p-auto-bump-unstable | u-libp2p-u-small-01.aws-eu-central-1a.nimbus.pyrmont

**Comment by jakubgs** [comment]
So the steps for `nim-libp2p-auto-bump-unstable` would be:

1. Checkout `unstable` in `nimbus-eth2`
2. Delete local version of `nim-libp2p-auto-bump-unstable`
3. Checkout new local `nim-libp2p-auto-bump-unstable` from `unstable` in `nimbus-eth2`
4. Bump the version of `vendor/nim-libp2p` to newest commit of `unstable` in `nim-libp2p` repo
5. Force push new version of `nim-libp2p-auto-bump-unstable`

@zah does that make sense?
Is there any specific way in which the bump of `vendor/nim-libp2p` has to be done?
Also, what job takes care of auto-bumping the `nim-libp2p-auto-bump-unstable` branch?

**Comment by jakubgs** [comment]
As far as I can tell the current state of `nimbus-eth2` branchs for `nim-libp2p-auto-bump` is wrong:

* The verison of `nim-libp2p` in `nim-libp2p-auto-bump` is `6b930ae`, which is the newest commit in `master`
* The version of `nim-libp2p` in `nim-libp2p-auto-bump-unstable` is `54031c9`, which is an older commit in `master`

Doesn't his mean that `nim-libp2p-auto-bump-unstable` is following the wrong branch in `nim-libp2p` @zah @dryajov ?

**Comment by dryajov** [comment]
> Doesn't his mean that nim-libp2p-auto-bump-unstable is following the wrong branch in nim-libp2p @zah @dryajov ?

I'll take a look at it.

> Also, what job takes care of auto-bumping the nim-libp2p-auto-bump-unstable branch?

Should be here - https://github.com/status-im/nim-libp2p/blob/master/.github/workflows/ci.yml#L209-L290

**Comment by dryajov** [comment]
OK, so nimbus now that https://github.com/status-im/nimbus-eth2/pull/2474 has been merged, nimbus should build against both latest libp2p master and unstable. I believe this should unblock this issue?

**Comment by dryajov** [comment]
The bump job also seems to have worked - https://github.com/status-im/nimbus-eth2/pull/2493

**Comment by arthurk** [comment]
can confirm that the build works now:

```
SUCCESS - pushed: statusteam/nimbus_beacon_node:unstable-libp2p-unstable-large
```

Next i'll distribute the validators so that we have 1250 running on libp2p-stable and 1250 on libp2p-unstable

**Comment by arthurk** [comment]
alright deployment is done

```
u-libp2p-s-large-01.aws-eu-central-1a.nimbus.prater
u-libp2p-u-large-01.aws-eu-central-1a.nimbus.prater
```

**Comment by dryajov** [comment]
Has `u-libp2p-s-large-01.aws-eu-central-1a.nimbus.prater` been also deployed, I can't seem to pull it's dashboard up - https://metrics.status.im/d/Xmoo2J-Gz/nimbus-fleet-libp2p-extras?orgId=1&from=now-24h&to=now&refresh=30s&var-instance=u-libp2p-s-large-01.aws-eu-central-1a.nimbus.prater

**Comment by jakubgs** [comment]
Hmm, the node is there:
```
admin@libp2p-large-01.aws-eu-central-1a.nimbus.prater:~ % d
CONTAINER ID   NAMES                                             IMAGE                                                        CREATED        STATUS
59753de036d3   beacon-node-prater-unstable-libp2p-stable-large   statusteam/nimbus_beacon_node:unstable-libp2p-stable-large   17 hours ago   Up 17 hours
```
And the consul definition looks correct:
```
admin@libp2p-large-01.aws-eu-central-1a.nimbus.prater:~ % jq '.services[] | select(.name=="beacon-node-metrics").address' /etc/consul/service_beacon_node.json     
"10.4.0.218"

```
I can see it in Consul:
https://consul.infra.status.im/ui/aws-eu-central-1a/nodes/libp2p-large-01.aws-eu-central-1a.nimbus.prater/health-checks

**Comment by jakubgs** [comment]
Ah, there it is:
https://metrics.status.im/d/Xmoo2J-Gz/nimbus-fleet-libp2p-extras?orgId=1&from=now-24h&to=now&refresh=30s&var-instance=libp2p-large-01.aws-eu-central-1a.nimbus.prater

The hostname was not corrected by running `infra-role-bootstrap`:
```
 > ssh admin@u-libp2p-s-large-01.aws-eu-central-1a.nimbus.prater hostname
libp2p-large-01.aws-eu-central-1a.nimbus.prater
```

**Comment by jakubgs** [comment]
There we go:
![image](https://user-images.githubusercontent.com/2212681/114677297-739f8500-9d0a-11eb-91fe-d4c2fdec8bda.png)

Is there anything left to do here? Do we care about renaming Pyrmont hosts? Or will we retire them soon enough?

**Comment by arthurk** [comment]
I guess this is done and can be closed

**Comment by jakubgs** [comment]
Since #45 will cover porting hosts to Hetzner I'm closing this.
